File: packages/types/src/checkpoint/agent/snapshot.ts
/**
 * Agent Loop State Snapshot Type Definition
 *
 * ## Design Principles
 *
 * 1. **Config NOT serialized**: `AgentLoopRuntimeConfig` contains callback functions
 *    (`transformContext`, `convertToLlm`) that cannot be serialized to JSON/TOML.
 *    Config must be re-provided by the application when restoring from checkpoint.
 *
 * 2. **Messages NOT in snapshot**: Messages are owned and managed by `ConversationSession`,
 *    not by `AgentLoopState`. The `AgentLoopStateSnapshot` captures only the execution
 *    progress state (iteration count, tool calls, status), while messages are handled
 *    through the conversation's checkpoint path.
 *
 * 3. **Restoration flow**: Application calls `AgentLoopFactory.fromCheckpoint()`
 *    with both the checkpoint ID and a fresh `AgentLoopRuntimeConfig`.
 *    The factory loads the snapshot, creates state from it, and injects the config.
 */

import { AgentLoopStatus } from "../../agent-execution/types.js";
import type { IterationRecord } from "../../agent-execution/types.js";
import type { ExecutionErrorRecord, ExecutionInterruptionRecord, ExecutionEventRecord } from "../execution-events.js";

/**
 * Agent Loop Status Snapshot
 *
 * Contains only serializable execution progress data.
 * Does NOT include:
 * - `config`: Must be re-provided by application (contains callbacks)
 * - `messages`: Managed by ConversationSession, not AgentLoopState
 *
 * ## Plan C Changes
 *
 * Added execution-related arrays to ensure all execution data is persisted
 * atomically with the state:
 * - `errors`: Errors that occurred during execution (was in ExecutionHistoryAPI)
 * - `interruptions`: Pauses/stops during execution (was in ExecutionHistoryAPI)
 * - `events`: Significant execution events for timeline tracking (was in ExecutionHistoryAPI)
 *
 * This eliminates the separate ExecutionHistoryAPI storage and ensures data
 * consistency and disaster recovery.
 */
export interface AgentLoopStateSnapshot {
  /** Execution status */
  status: AgentLoopStatus;
  /** Current iteration number */
  currentIteration: number;
  /** Total tool call count */
  toolCallCount: number;
  /** Execution start timestamp (ms) */
  startTime: number | null;
  /** Execution end timestamp (ms) */
  endTime: number | null;
  /** Error data (if execution failed) - DEPRECATED: use errors array */
  error: unknown;

  // ========== Plan C: Execution Event Tracking ==========

  /** Errors that occurred during execution (atomic with state) */
  errorRecords?: ExecutionErrorRecord[];

  /** Interruptions (pauses/stops) that occurred during execution */
  interruptionRecords?: ExecutionInterruptionRecord[];

  /** Recent execution events for timeline view (limited to prevent state bloat) */
  eventRecords?: ExecutionEventRecord[];

  // ========== Extended fields for complete state capture ==========

  /** Iteration history records */
  iterationHistory?: IterationRecord[];
  /** Current in-progress iteration record (if any) */
  currentIterationRecord?: IterationRecord;

  /** Dynamic tools (serialized from Set<string> to string[]) */
  dynamicTools?: string[];

  // ========== Streaming state fields (for pause/resume precision) ==========

  /** Whether the agent was streaming an LLM response at checkpoint time */
  isStreaming?: boolean;
  /** Incomplete streaming message content (for stream resume) */
  streamMessage?: unknown;
  /** Pending tool call IDs that were in-flight at checkpoint time */
  pendingToolCallIds?: string[];

  // ========== Trigger state (for tracking trigger fires and limits) ==========

  /** Trigger state snapshot (serialized from TriggerStateManager) */
  triggerState?: Record<string, {
    triggerId: string;
    fireCount: number;
    lastFiredAt?: number;
    metadata?: Record<string, unknown>;
  }>;

  // ========== Variable history tracking ==========

  /** Variable snapshots captured at each iteration for debugging */
  variableSnapshots?: Array<{
    /** Iteration number when snapshot was taken */
    iteration: number;
    /** Timestamp when snapshot was taken */
    timestamp: number;
    /** Variables at this point in execution */
    variables: Record<string, VariableSnapshot>;
    /** Tool call ID if snapshot was taken after a tool call */
    toolCallId?: string;
  }>;
}

/**
 * Variable snapshot details
 */
export interface VariableSnapshot {
  /** Variable value at this point */
  value: unknown;
  /** Type of the value */
  type: string;
  /** Size in bytes (for large objects) */
  size?: number;
  /** Whether this variable was updated in this iteration */
  updated: boolean;
  /** Source of the variable (user, tool, system) */
  source: 'user' | 'tool' | 'system';
}

File: crates/wf-types/src/checkpoint/agent/snapshot.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::execution::ExecutionHierarchy;
use crate::message::Message;
use crate::Id;
use crate::Timestamp;

/// Agent loop state snapshot aligned with the TS `AgentLoopStateSnapshot`.
///
/// Optional fields mirror the TS snapshot: absent fields are simply skipped
/// during capture or restore.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AgentStateSnapshot {
    pub agent_loop_id: Id,
    pub status: String,
    pub current_iteration: u32,
    pub tool_call_count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_snapshot: Option<Vec<Message>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_history: Option<Vec<serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_streaming: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub variable_snapshots: Option<HashMap<String, VariableSnapshot>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub started_at: Option<Timestamp>,
    pub completed_at: Option<Timestamp>,
    /// Execution error records (Plan C).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_records: Option<Vec<serde_json::Value>>,
    /// Execution interruption records (Plan C).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interruption_records: Option<Vec<serde_json::Value>>,
    /// Execution event records (Plan C).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event_records: Option<Vec<serde_json::Value>>,
    /// History of completed iterations.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub iteration_history: Option<Vec<serde_json::Value>>,
    /// The iteration record currently being executed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub current_iteration_record: Option<serde_json::Value>,
    /// Streaming message content in flight.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_message: Option<String>,
    /// Tool call ids waiting for results.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pending_tool_call_ids: Option<Vec<String>>,
    /// Trigger state (trigger fires / limits).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trigger_state: Option<serde_json::Value>,
    /// Execution hierarchy metadata (children references).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hierarchy: Option<ExecutionHierarchy>,
    /// Messages from the conversation session (includeMessages content config).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub messages: Option<Vec<Message>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct VariableSnapshot {
    pub value: serde_json::Value,
    pub r#type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size: Option<u64>,
    /// Whether the variable was updated since the previous checkpoint
    /// (aligned with the TS `VariableSnapshot.updated: boolean`).
    pub updated: bool,
    pub source: String,
}
